pkg/controller/user-cluster-controller-manager/ccm-csi-migrator/controller.go

```diff
@@ -0,0 +1,173 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "ccm-csi-migrator"
+)
+
+type reconciler struct {
+	log          *zap.SugaredLogger
+	seedClient   ctrlruntimeclient.Client
+	userClient   ctrlruntimeclient.Client
+	seedRecorder record.EventRecorder
+	versions     kubermatic.Versions
+	clusterName  string
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, seedMgr, userMgr manager.Manager, versions kubermatic.Versions, clusterName string) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		log:          log,
+		seedClient:   seedMgr.GetClient(),
+		userClient:   userMgr.GetClient(),
+		seedRecorder: seedMgr.GetEventRecorderFor(controllerName),
+		versions:     versions,
+		clusterName:  clusterName,
+	}
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller %s: %v", controllerName, err)
+	}
+
+	// Watch for changes to Machines
+	if err = c.Watch(
+		&source.Kind{Type: &v1alpha1.Machine{}},
+		handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: clusterName,
+					},
+				},
+			}
+		}),
+	); err != nil {
+		return fmt.Errorf("failed to establish watch for the Machines %v", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("Request", request.NamespacedName.String())
+	log.Debug("Reconciling")
+
+	cluster := &v1.Cluster{}
+	if err := r.seedClient.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debug("cluster not found, returning")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get cluster: %v", err)
+	}
+
+	err := r.reconcile(ctx, log, cluster)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.seedRecorder.Event(cluster, corev1.EventTypeWarning, "CCMCSIMigrationFailed", err.Error())
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, oldCluster *v1.Cluster) error {
+	machines := &v1alpha1.MachineList{}
+	if err := r.userClient.List(ctx, machines); err != nil {
+		log.Debugw("error while listing machines in user oldCluster", zap.Error(err))
+		return err
+	}
+
+	// check all the machines have been migrated
+	var migrated = true
+	for _, machine := range machines.Items {
+		flag := common.GetKubeletFlags(&machine)[common.ExternalCloudProviderKubeletFlag]
```

Comment:
Does it make sense to also check for the CSI migration feature gates?

Reply:
I think that checking also CSI migration feature gates would add an unneeded complexity since all the providers have different feature gates, and we should distinguish between them. Of course, we can do it, but tbh I don't know if it worths, since the CSI feature flags are added along to the `ExternalCloudProviderKubeletFlag`: https://github.com/kubermatic/machine-controller/blob/3a115014a952d06d867ecee5e8d780acafefc289/pkg/admission/machines.go#L87

Reply:
> I think that checking also CSI migration feature gates would add an unneeded complexity since all the providers have different feature gates, and we should distinguish between them.

While this is true, there's also the `CSIMigration` feature gate that's common for all cloud providers. However, it increases the complexity a little bit, so I'm also not sure is it really worth it. I don't think we'll ever have a situation that Machine might have the external cloud provider enabled, but not the CSIMigration feature gates (for migrated clusters).

I think that we can ignore this for now.